api/src/main/java/org/pmiops/workbench/initialcredits/InitialCreditsExpirationServiceImpl.java

```diff
@@ -1,14 +1,64 @@
 package org.pmiops.workbench.initialcredits;
 
+import jakarta.inject.Provider;
+import jakarta.mail.MessagingException;
 import java.sql.Timestamp;
+import java.time.Clock;
+import java.util.List;
 import java.util.Optional;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.db.dao.UserDao;
+import org.pmiops.workbench.db.dao.WorkspaceDao;
 import org.pmiops.workbench.db.model.DbUser;
 import org.pmiops.workbench.db.model.DbUserInitialCreditsExpiration;
+import org.pmiops.workbench.db.model.DbUserInitialCreditsExpiration.NotificationStatus;
+import org.pmiops.workbench.db.model.DbWorkspace;
+import org.pmiops.workbench.exceptions.WorkbenchException;
+import org.pmiops.workbench.leonardo.LeonardoApiClient;
+import org.pmiops.workbench.mail.MailService;
+import org.pmiops.workbench.model.BillingStatus;
+import org.pmiops.workbench.workspaces.WorkspaceUtils;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
 @Service
 public class InitialCreditsExpirationServiceImpl implements InitialCreditsExpirationService {
 
+  private static final org.slf4j.Logger logger =
+      LoggerFactory.getLogger(InitialCreditsExpirationServiceImpl.class);
+  private final UserDao userDao;
+  private final MailService mailService;
+  private final Clock clock;
+  private final WorkspaceDao workspaceDao;
+  private final Provider<WorkbenchConfig> workbenchConfigProvider;
+  private final LeonardoApiClient leonardoApiClient;
+
+  @Autowired
+  public InitialCreditsExpirationServiceImpl(
+      UserDao userDao,
+      MailService mailService,
+      WorkspaceDao workspaceDao,
+      Clock clock,
+      Provider<WorkbenchConfig> workbenchConfigProvider,
+      LeonardoApiClient leonardoApiClient) {
+    this.userDao = userDao;
+    this.mailService = mailService;
+    this.clock = clock;
+    this.workspaceDao = workspaceDao;
+    this.workbenchConfigProvider = workbenchConfigProvider;
+    this.leonardoApiClient = leonardoApiClient;
+  }
+
+  @Override
+  public void checkCreditsExpirationForUserIDs(List<Long> userIdsList) {
+    if (userIdsList != null && !userIdsList.isEmpty()) {
+      Timestamp now = new Timestamp(clock.instant().toEpochMilli());
```

Comment:
A general feedback on this PR: 
You should avoid using the date/time APIs in java.util. This includes java.util.Date (and its subclass, java.sql.Timestamp), java.util.Calendar, java.util.TimeZone, etc. They are [bad APIs](https://dzone.com/articles/game-over-jdks-date-and-time-c). Even Oracle now [acknowledges](https://docs.oracle.com/javase/tutorial/datetime/iso/legacy.html) that these classes have "several drawbacks"